### PR TITLE
CROSS join ignore join_algorithm setting

### LINKED ISSUE #62431: Seems the new anaylzer has something wrong under 'parallel_hash' join algorithm
If we launch a clickhouse server with `<join_algorithm>parallel_hash</join_algorithm>` in the config file, we will get an exception from client command:
`ClickHouse client version 24.3.2.1.
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 24.3.2.
Cannot load data for command line suggestions: Code: 48. DB::Exception: Received from localhost:9000. DB::Exception: Only hash join supports multiple ORs for keys in JOIN ON section. (NOT_IMPLEMENTED) (version 24.3.2.1)
`

Here is the SQL that triggered the problem and it is from clickhouse client:
`SELECT DSELECT DISTINCT arrayJoin(extractAll(name, '[\\w_]{2,}')) AS res FROM (SELECT * FROM viewIfPermitted(SELECT name FROM system.functions ELSE null('name String')) UNION ALL SELECT * FROM viewIfPermitted(SELECT name FROM system.table_engines ELSE null('name String')) UNION ALL SELECT * FROM viewIfPermitted(SELECT name FROM system.formats ELSE null('name String')) UNION ALL SELECT * FROM viewIfPermitted(SELECT name FROM system.table_functions ELSE null('name String')) UNION ALL SELECT * FROM viewIfPermitted(SELECT name FROM system.data_type_families ELSE null('name String')) UNION ALL SELECT * FROM viewIfPermitted(SELECT name FROM system.merge_tree_settings ELSE null('name String')) UNION ALL SELECT * FROM viewIfPermitted(SELECT name FROM system.settings ELSE null('name String')) UNION ALL SELECT * FROM viewIfPermitted(SELECT cluster FROM system.clusters ELSE null('cluster String')) UNION ALL SELECT * FROM viewIfPermitted(SELECT macro FROM system.macros ELSE null('macro String')) UNION ALL SELECT * FROM viewIfPermitted(SELECT policy_name FROM system.storage_policies ELSE null('policy_name String')) UNION ALL SELECT * FROM viewIfPermitted(SELECT concat(func.name, comb.name) AS x FROM system.functions AS func CROSS JOIN system.aggregate_function_combinators AS comb WHERE is_aggregate ELSE null('x String')) UNION ALL SELECT * FROM viewIfPermitted(SELECT name FROM system.databases LIMIT 10000 ELSE null('name String')) UNION ALL SELECT * FROM viewIfPermitted(SELECT DISTINCT name FROM system.tables LIMIT 10000 ELSE null('name String')) UNION ALL SELECT * FROM viewIfPermitted(SELECT DISTINCT name FROM system.dictionaries LIMIT 10000 ELSE null('name String')) UNION ALL SELECT * FROM viewIfPermitted(SELECT DISTINCT name FROM system.columns LIMIT 10000 ELSE null('name String'))) WHERE notEmpty(res)`

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* CROSS join can be executed with any value `join_algorithm` setting, close https://github.com/ClickHouse/ClickHouse/issues/62431

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>Modify your CI run</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  All with Azure
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

<details>

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit b8806dcb6572c54c82c863c7636a3006a9193461 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63273/b8806dcb6572c54c82c863c7636a3006a9193461/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Planner/PlannerJoins.cpp [hunk 1]
```diff
@@ -878,6 +878,9 @@ std::shared_ptr<IJoin> chooseJoinAlgorithm(std::shared_ptr<TableJoin> & table_jo
         return std::make_shared<HashJoin>(table_join, right_table_expression_header);
     }
 
+    if (table_join->kind() == JoinKind::Cross)
```
Comment (novikd) [inline src/Planner/PlannerJoins.cpp:881]: Requires a comment. The logic is unclear.

File: tests/queries/0_stateless/02815_join_algorithm_setting.sql [hunk 2]
```diff
@@ -94,3 +94,13 @@ SELECT countIf(explain like '%Algorithm: DirectKeyValueJoin%'), countIf(explain
...
 );
+
+SET join_algorithm = 'grace_hash';
+
+-- Cannot execute the grace hash with OR condition
+SELECT * FROM ( SELECT number AS key, number * 10 AS key2 FROM numbers_mt(10) ) AS t1
+JOIN ( SELECT k AS key, k + 100 AS key2 FROM t2 ) AS t2 ON t1.key = t2.key OR t1.key2 = t2.key2; -- { serverError NOT_IMPLEMENTED }
+
+-- But for CROSS choose `hash` algorithm even though it's not enabled
+SELECT * FROM ( SELECT number AS key, number * 10 AS key2 FROM numbers_mt(10) ) AS t1
```
Comment (novikd) [inline tests/queries/0_stateless/02815_join_algorithm_setting.sql:105]: This query fails with exception:
```
Can't execute any of specified join algorithms for this strictness/kind and right storage type. (NOT_IMPLEMENTED)
```

Comment (vdimir) [reply tests/queries/0_stateless/02815_join_algorithm_setting.sql:105]: For old analyzer behavior remains, added setting to the test

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**